### PR TITLE
Added getPledgesByUserId Resolver & added pledges and campaigns in AppUserProfile

### LINKED ISSUE #2431: GSoC: Add Resolvers for getting user-linked Pledges and Campaigns
**Is your feature request related to a problem? Please describe.**
Following issue is being addressed in GSoC proj:
In Talawa Admin, we require to get User linked pledges. We require resolvers to get these with sort and search support for available attributes. The following feature is required for adding campaign and pledge screen in Talawa User Portal.

**Describe the solution you'd like**
- Add pledges[] and campaigns[] in `AppUserProfile` so as to easily get pledges and campaigns linked to a user.
- Update all the related resolvers (`createFundraisingCampaignPledge`, `removeFundraisingCampaignPledge`, `updateFundraisingCampaignPledge`, etc..) to take in account pledges[] & campaigns[] in AppUserprofile 
- Proposed Resolvers:

1. `getPledgesByUserId(orderBy: PledgeOrderByInput, userId: ID!, where: PledgeWhereInput): [FundraisingCampaignPledge]`
2. `getFundraisingCampaigns(campaignOrderby: CampaignOrderByInput, pledgeOrderBy: PledgeOrderByInput, where: CampaignWhereInput): [FundraisingCampaign]!`

**Additional context**
Following Feature needs to be added in order to support [Talawa Admin Issue](https://github.com/PalisadoesFoundation/talawa-admin/issues/2070)

### PR DESCRIPTION
**Pull request by GlenDsza**
**What kind of change does this PR introduce?**

bugfix, feature, refactoring

**Issue Number:**

Fixes #2431 

**Did you add tests for your changes?**

Yes

**Summary**

- Added `getPledgesByUserId`  query resolver
- Added pledges & campaigns attribute in AppUserProfile Model implement efficient searching of pledges and campaigns for Users
- Modified Create/Remove/Update mutations of Fund, Campaign, Pledges taking in consideration addition of pledges and campaigns attribute in AppUserProfile.
- Fixed bug in login which lead to creation of duplicate AppUserProfile for superadmin on every login.
- Changes campaigns attribute in Pledge Model to campaign as every pledge is related to a single campaign
- Added tests for above changes.

**Does this PR introduce a breaking change?**

Yes, 
- Few screens related to Fund/Campaign/Pledges in Talawa Admin might break due to changes in query
- To handle this, before merging this API PR review PR related to following issue in [Link](https://github.com/PalisadoesFoundation/talawa-admin/issues/2070) Talawa Admin
- Then merge UI PR once this is merged

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

- **New Features**
  - Enhanced GraphQL schema with new fields for user profiles and campaign inputs.
  - Introduced new queries for fetching fundraising campaigns and user-specific pledges.

- **Bug Fixes**
  - Improved pledge management during campaign removal to ensure data integrity.

- **Documentation**
  - Updated comments and documentation to reflect new properties and functionality across various data models.

- **Tests**
  - Added comprehensive test coverage for new features and changed behaviors, ensuring proper validation of campaign and pledge associations.

### CODE REVIEW

**Comment by palisadoes** [comment]
@SiddheshKukade @tasneemkoushar Please review

File: src/resolvers/Query/getPledgesByUserId.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  // Filter the pledges based on pledger's name
+  if (args.where?.firstName_contains) {
+    appUserProfile.pledges = appUserProfile.pledges.filter((pledge) => {
+      const tempPledge = pledge as InterfaceFundraisingCampaignPledges;
+      const user = tempPledge.users[0] as InterfaceUser;
+      return user.firstName.includes(args?.where?.firstName_contains as string);
+    });
+  }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Query/getPledgesByUserId.ts:62]: As both filters are separate and cant be applied simultaneously, clubbing them wont make it efficient

File: tests/helpers/FundraisingCampaign.ts [hunk 2]
```diff
@@ -11,16 +11,20 @@
...
   fundId: string,
+  organizationId?: string | undefined,
+  populated?: string[],
 ): Promise<InterfaceFundraisingCampaign> => {
   //   const [testUser, testOrganization, testFund] = await createTestFund();
 
   const testFundraisingCampaign = await FundraisingCampaign.create({
     name: `name${nanoid().toLowerCase()}`,
     fundId: fundId,
+    organizationId: organizationId ?? "organizationId",
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/helpers/FundraisingCampaign.ts]: Done

File: src/resolvers/Mutation/removeFundraisingCampaign.ts [hunk 3]
```diff
@@ -122,6 +126,23 @@ export const removeFundraisingCampaign: MutationResolvers["removeFundraisingCamp
...
+    const pledgesTobeDeleted: Types.ObjectId[] = [];
+    for (const pledge of campaign.pledges as InterfaceFundraisingCampaignPledges[]) {
+      // Remove pledges & campaign from related user's AppUserProfile
+      pledgesTobeDeleted.push(pledge._id);
+
+      await AppUserProfile.updateMany(
+        { userId: { $in: pledge.users } },
+        { $pull: { pledges: pledge._id, campaigns: campaign._id } },
+      );
+    }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/removeFundraisingCampaign.ts:139]: Done

File: src/resolvers/Mutation/updateFundCampaignPledge.ts [hunk 4]
```diff
@@ -82,6 +83,52 @@ export const updateFundraisingCampaignPledge: MutationResolvers["updateFundraisi
...
+
+      // Update AppUserProfile for every user who is newly added to the pledge
+      await AppUserProfile.updateMany(
+        {
+          userId: { $in: usersAdded },
+        },
+        {
+          $addToSet: { pledges: pledge._id, campaigns: pledge.campaign },
+        },
+      );
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/updateFundCampaignPledge.ts:131]: Done

File: src/resolvers/Mutation/updateFundCampaignPledge.ts [hunk 5]
```diff
@@ -82,6 +83,52 @@
...
+        );
+
+        if (otherPledges.length === 0) {
+          await AppUserProfile.findOneAndUpdate(
+            { userId },
+            { $pull: { campaigns: campaignId } },
+            { new: true },
+          );
+        }
+      }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/updateFundCampaignPledge.ts:116]: Done

**Comment by SiddheshKukade** [comment]
Checking

**Review by tasneemkoushar — APPROVED** [review]

**Comment by tasneemkoushar** [comment]
@GlenDsza LGTM. Feel free to merge

**Pull request merged**